index.js

```diff
@@ -18,149 +10,197 @@ var mkdirp = require('mkdirp');
- * @return {undefined}
+ * @name .stream
+ * @param {String} `filepath` file path.
+ * @param {Object} `options` Options to pass to [fs.createWriteStream][wsoptions]
+ * @return {Stream} Returns a new [WriteStream][writestream] object. (See [Writable Stream][writable]).
  * @api public
  */
 
-writeFile.sync = function(filepath, data, options) {
+write.stream = (filepath, contents, options) => {
   if (typeof filepath !== 'string') {
     throw new TypeError('expected filepath to be a string');
   }
-  mkdirp.sync(path.dirname(filepath), options);
-  fs.writeFileSync(filepath, data, options);
+
+  const opts = { encoding: 'utf8', ...options };
+  const destpath = opts.increment ? incrementName(filepath) : filepath;
+
+  if (opts.overwrite === false && exists(filepath, destpath)) {
+    throw new Error('File already exists: ' + filepath);
+  }
+
+  mkdirSync(path.dirname(destpath), { recursive: true, ...options });
+  return fs.createWriteStream(destpath, opts);
 };
 
 /**
- * Uses `fs.createWriteStream` to write data to a file, replacing the
- * file if it already exists and creating any intermediate directories
- * if they don't already exist. Data can be a string or a buffer. Returns
- * a new [WriteStream](https://nodejs.org/api/fs.html#fs_class_fs_writestream)
- * object.
- *
- * ```js
- * var fs = require('fs');
- * var writeFile = require('write');
- * fs.createReadStream('README.md')
- *   .pipe(writeFile.stream('a/b/c/other-file.md'))
- *   .on('close', function() {
- *     // do stuff
- *   });
- * ```
- * @name .stream
- * @param {string|Buffer|integer} `filepath` filepath or file descriptor.
- * @param {object} `options` Options to pass to [mkdirp][] and [fs.createWriteStream][fs]{#fs_fs_createwritestream_path_options}
- * @return {Stream} Returns a new [WriteStream](https://nodejs.org/api/fs.html#fs_class_fs_writestream) object. (See [Writable Stream](https://nodejs.org/api/stream.html#stream_class_stream_writable)).
- * @api public
+ * Increment the filename if the file already exists and enabled by the user
  */
 
-writeFile.stream = function(filepath, options) {
-  mkdirp.sync(path.dirname(filepath), options);
-  return fs.createWriteStream(filepath, options);
+const incrementName = destpath => {
+  let file = { ...path.parse(destpath), path: destpath };
+  let name = file.name;
+  let n = 1;
+
+  while (fs.existsSync(file.path)) {
+    file.path = path.join(file.dir, `${name} (${++n})${file.ext}`);
+  }
+
+  return file.path;
+};
+
+/**
+ * Ensure newline at EOF if defined on options
+ */
+
+const ensureNewline = (data, options) => {
+  if (!options || options.newline !== true) return data;
+  if (typeof data !== 'string' && !isBuffer(data)) {
+    return data;
+  }
+
+  // only call `.toString()` on the last character. This way, if
+  // data is a buffer, we only need to call `.toString()` on
+  // the entire string if the condition is true.
+  if (String(data.slice(-1)) !== '\n') {
+    return data.toString() + '\n';
```

Comment:
It should be possible to append a newline to the buffer without doing `.toString()`.

Reply:
nice, good catch! I wasn't thinking about that. This is an awesome example of a place where we can make code much more performant with a really simple optimization. Awesome.

Reply:
I changed it to:

```js
  if (String(data.slice(-1)) !== '\n') {
    return data.concat(Buffer.from('\n'));
  }
```

Reply:
I think this should be possible with other things too, like front matter. I'll look into it.

Reply:
@doowb I just saw your point below about the Buffer class. LMKWYT.